Feature: in-process benchmarks

ig-sinicyn opened the pull request:
This PR adds toolchain that enables in-process benchmark running.

The main benefit of it is ability to run a lot of benchmarks very quickly in a cost of less accurate results.

In practice we're getting stable results with 2-3 seconds per benchmark time limit (fast enough if compared to typical 20-60 secs for outofproc toolchain).
*Note: we use custom config like this:
```c#
		public static readonly Job DefaultJob = new Job("Competition")
		{
			Env =
			{
				Gc =
				{
					Force = false
				}
			},
			Run =
			{
				LaunchCount = 1,
				WarmupCount = 100,
				TargetCount = 300,
				RunStrategy = RunStrategy.Throughput,
				UnrollFactor = 16,
				InvocationCount = 256
			},
			Infrastructure =
			{
				Toolchain = InProcessToolchain.DontLogOutput
			}
		}.Freeze();
```

The implementation is in sync of content of `BenchmarkProgram.txt` so with same setup results are expected to be pretty close to results obtained with default toolchain (and they are).

By default toolchain uses Reflection.Emit to emit unrolled body of the loops but in case the benchmark run in platform that does not support dynamic code generation (e.g. .net native)
there's a fallback code that uses Delegate.Combine to unroll the loop (can be set via BenchmarkActionCodegen enum).
Because of the latter, the body of the Dummy method (BenchmarkActionFactory.DummyInstance) is hardcoded.

Also, there's an option that suppress output from the toolshain's executor. It makes the log much more shorter:)

Finally, there is a InProcessValidator that helps to ensure that surrent envirement matches to the Job settings.

More details and things to improve - at #358.

dnfclas commented:
@ig-sinicyn,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

AndreyAkinshin commented:
@ig-sinicyn, could you add an example of an in-process benchmark in the `BenchmarkDotNet.Samples` project? (For example, `IntroInProcess`)

ig-sinicyn commented:
@AndreyAkinshin yes, of course.

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessExecutor.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <summary> Default timeout for in-process benchmarks. </summary>
+        public static readonly TimeSpan DefaultTimeout = TimeSpan.FromMinutes(5);
+
+        #region .ctor & properties
```
nit: we don't use regions in BenchmarkDotNet, they are ugly ;)

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessExecutor.cs:
```diff
@@ -0,0 +1,150 @@
+            var runThread = new Thread(() => exitCode = ExecuteCore(host, benchmark, logger));
+            if (benchmark.Target.Method.GetCustomAttributes<STAThreadAttribute>(false).Any())
+            {
+                // TODO: runThread.SetApartmentState(ApartmentState.STA);
```
it's great that you remember that we support STA, may I ask why is this commented out?

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessExecutor.cs:
```diff
@@ -0,0 +1,150 @@
+            BuildResult buildResult, Benchmark benchmark, ILogger logger, IResolver resolver, IDiagnoser diagnoser = null)
+        {
+            // TODO: preallocate buffer for output (no direct logging)?
+            var host = new InProcessHost(benchmark, LogOutput ? logger : null, diagnoser);
```
maybe instead of passing null you could implement some Null Object pattern here (logger that does nothing)?

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessExecutor.cs:
```diff
@@ -0,0 +1,150 @@
+            var timeout = HostEnvironmentInfo.GetCurrent().HasAttachedDebugger ? UnderDebuggerTimeout : ExecutionTimeout;
+
+            runThread.Start();
+            // TODO: to timespan overload (not available for .Net Core for now).
```
you might use conditional switches if something is not available for .NET Core
#if !CORE
#else
#endif

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessExecutor.cs:
```diff
@@ -0,0 +1,150 @@
+                {
+                    process.TrySetAffinity(affinity.Value, logger);
+                }
+                // TODO: set thread priority to highest
```
why setting the priority was left as todo? is there some contract mismatch (.net core and stuff)?

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessExecutor.cs:
```diff
@@ -0,0 +1,150 @@
+            {
+                lines.Add(measurement.ToOutputLine());
+            }
+            lines.Add(runResults.GCStats.WithTotalOperations(runResults.TotalOperationsCount).ToOutputLine());
```
👍 thanks for keeping the MemoryDiagnoser support!

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessRunner.cs:
```diff
@@ -0,0 +1,137 @@
+
+            try
+            {
+                // we are not using Runnable here in any direct way in order to avoid strong dependency Main<=>Runnable
```
👍 thanks for keeping the old behaviour!

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessRunner.cs:
```diff
@@ -0,0 +1,137 @@
+                setupAction.InvokeSingle();
+
+                if (job.ResolveValue(RunMode.RunStrategyCharacteristic, EngineResolver.Instance).NeedsJitting())
+                    engine.Jitting(); // does first call to main action, must be executed after setup()!
```
👍 thanks for keeping another detail in mind!

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessRunner.cs:
```diff
@@ -0,0 +1,137 @@
+                    flags |= parameter.IsStatic ? BindingFlags.Static : BindingFlags.Instance;
+
+                    var targetType = benchmark.Target.Type;
+                    var paramProperty = targetType.GetProperty(parameter.Name, flags);
```
does this method support public fields? (not only properties)

ig-sinicyn commented:
@AndreyAkinshin 
Done.

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessToolchain.cs:
```diff
@@ -0,0 +1,69 @@
+        public static readonly IToolchain Instance = new InProcessToolchain(true);
+
+        /// <summary>The toolchain instance without output logging.</summary>
+        public static readonly IToolchain DontLogOutput = new InProcessToolchain(false);
```
👍 that's a very nice and desired feature!

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessToolchain.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <param name="logger">The logger.</param>
+        /// <param name="resolver">The resolver.</param>
+        /// <returns><c>true</c> if the benchmark can be run with the toolchain.</returns>
+        public bool IsSupported(Benchmark benchmark, ILogger logger, IResolver resolver) => true;
```
I believe that there are few scenarios that you can not support. For example: architecture/jit different than host/current

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessValidator.cs:
```diff
@@ -0,0 +1,145 @@
+
+        // ReSharper restore HeapView.DelegateAllocation
+
+        private static string DontValidate(Job job, Characteristic characteristic) => null;
```
👍 very smart way to show that you don't validate something on purpose!

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/[BenchmnarkActions]/BenchmarkActionFactory.cs:
```diff
@@ -0,0 +1,138 @@
+                        resultInstance, targetMethod,
+                        codegenMode, unrollFactor);
+
+                if (typeof(ValueTask<>).IsAssignableFrom(genericType))
```
👍 thanks for keeping another nasty detail in mind!

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Extensions/ProcessExtensions.cs:
```diff
@@ -25,5 +26,55 @@ public static void EnsureProcessorAffinity(this Process process, IntPtr value)
+            try
+            {
+                var cpuMask = (1 << Environment.ProcessorCount) - 1;
+                process.ProcessorAffinity = new IntPtr(processorAffinity.ToInt32() & cpuMask);
```
Nice trick with `cpuMask`! Could we introduce a helper function and use it in `EnsureProcessorAffinity`?

ig-sinicyn replied:
Yep, will add.

ig-sinicyn replied:
Added. Also, fixed dumb mistake with processorAffinity.ToInt32().

adamsitnik commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/[BenchmnarkActions]/BenchmarkActionFactory_Base.cs:
```diff
@@ -0,0 +1,228 @@
+                if (unrollFactor <= 1)
+                    return callback;
+
+                return (TDelegate)(object)Delegate.Combine(
```
👍 another smart idea!

adamsitnik commented on tests/BenchmarkDotNet.IntegrationTests/InProcessTest.cs:
```diff
@@ -0,0 +1,388 @@
+        [Fact]
+        public void TestBenchmarkActionRefType() => TestInvoke(x => x.InvokeOnceRefType(), UnrollFactor, StringResult);
+        [Fact]
+        public void TestBenchmarkActionValueType() => TestInvoke(x => x.InvokeOnceValueType(), UnrollFactor, DecimalResult);
```
personally, I don't like to use [Facts] and not following the naming convention. For example `TestBenchmarkActionValueType` is not a fact, `BenchmarksReturningValueTypesAreSupported` is

adamsitnik commented on tests/BenchmarkDotNet.IntegrationTests/InProcessTest.cs:
```diff
@@ -0,0 +1,388 @@
+                var summary = CanExecute<BenchmarkAllocates>(config);
+
+                var testLog = logger.GetLog();
+                Assert.Contains("// Benchmark: BenchmarkAllocates.Allocates:", testLog);
```
it would be better if you could use the strong-typed properties of the summary instead of looking for strings in the output

AndreyAkinshin commented:
@ig-sinicyn, please don't use square brackets in folder names (`[BenchmarkActions]`).

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Toolchains/InProcess/InProcessHost.cs:
```diff
@@ -0,0 +1,115 @@
+        public InProcessHost([NotNull] Benchmark benchmark, [CanBeNull] ILogger logger, [CanBeNull] IDiagnoser diagnoser)
+        {
+            if (benchmark == null)
+                throw new ArgumentNullException(nameof(benchmark));
```
Remove this check or replace `[NotNull]` by `[CanBeNull]`.

ig-sinicyn commented:
@adamsitnik 

Wow!:)

Summarized to keep the discussion simple :)

0. **what is the benefit of using the mask here?**
Process.ProcessorAffinity setter throws if the value is larger than avalable processors mask :)
I've learned it the hard way:)

1. **XML comments**: ported as is, sorry:)
Should I remove these 'dummy' comments?
If yes - all of them?
Explanation: we're using sandcastle to build helpfile on each build, so it's much easier to add a stub for help than re-comment all code later.

2. **#regions** - ok, they are gone:)

3. **InProcessHost:** logger that does nothing)?
+1, replaced with empty CompositeLogger.

4. **STA, why is this commented out?** and **why setting the priority was left as todo?**
I'm afraid .Net core has no API for it:(

5. **InProcessRunner.FillProperties:** 
nice catch! Fields are supported now.

6. **Toolchain.IsSupported:** 
this is already checked via in-process validator:)
Should I call the validator from InProcessToolchain.IsSupported?

7. **Personally, I don't like to use [Facts]** 
Fixed.

8. **it would be better if you could use the strong-typed properties of the summary** 
They are checked too. 
```c#
Assert.True(summary.Reports.Sum(r => r.GcStats.AllocatedBytes) > 0);
```
The output is checked to ensure that it is not broken:)

Will fix (2) and (6) as answered, others are fixed.


@AndreyAkinshin yours are not done, will commit later.

adamsitnik commented:
@ig-sinicyn thanks for applying the fixes and providing the answers! 

> Should I remove these 'dummy' comments?

No, but please don't add more in the future.

> STA, why is this commented out? and why setting the priority was left as todo?

I think that it would be better to do what we can for the classic .NET like
```cs
#if CLASSIC
setSta()
#endif
```

> this is already checked via in-process validator:)
> Should I call the validator from InProcessToolchain.IsSupported?

Thanks for the explanation. Is it possible to make sure that if user is using `InProcessToolchain` the `InProcessValidator` is always enabled? You did a great job with this validator, I just want to make sure people always get clear message what is possible and what is not.

btw do you have a twitter account so I can mention you in some conversations?

adamsitnik commented:
@ig-sinicyn the build has failed, can you take a look?

Otherwise LGTM and ready to merge!

ig-sinicyn commented:
@adamsitnik 

Sorry for broken build, missed a file. Fixed.

---

> No, but please don't add more in the future.

Ok, I've got it:)

>I think that it would be better to do what we can for the classic .NET like

Done.

>Thanks for the explanation. Is it possible to make sure that if user is using InProcessToolchain the InProcessValidator is always enabled?

Ok, there's what we can do:

1. Leave as is - no warnings, it silently works even if the environment do not match config.

2. Remove the validator and move the validator logic into `Toolchain.IsSuported()` - ugly. Validation errors will not be exposed by `Summary.ValidationErrors`.

3. Call validator from toolchain and log validation results as verbose (with `// ` prefix). Allow users to add validators if they want explicit validation.

I'm going to do the third. Better ideas, anyone? :)

>btw do you have a twitter account so I can mention you in some conversations?

Very old one https://twitter.com/Sinix . I almost do not use it, but will answer if there will be any questions.

ig-sinicyn commented:
@adamsitnik 

> Is it possible to make sure that if user is using InProcessToolchain the InProcessValidator is always enabled?

Done (see prev post for details).

AndreyAkinshin commented:
@adamsitnik  I think that you hurry to do it, I still have some concerns. The biggest one: here is my output for `IntroInProcess`:
```
// * Detailed results *
IntroInProcess.'new byte[10kB]': InProcess(Toolchain=InProcessToolchain, LaunchCount=1, TargetCount=15, WarmupCount=10)
Runtime = ; GC =
Mean = 533.4784 ns, StdErr = 6.2640 ns (1.17%); N = 15, StdDev = 24.2602 ns
Min = 503.6674 ns, Q1 = 511.8996 ns, Median = 530.5997 ns, Q3 = 548.4981 ns, Max = 587.9064 ns
IQR = 36.5985 ns, LowerFence = 457.0019 ns, UpperFence = 603.3958 ns
ConfidenceInterval = [521.2010 ns; 545.7558 ns] (CI 95%)
Skewness = 0.74, Kurtosis = 2.44


IntroInProcess.'stackalloc byte[10kB]': InProcess(Toolchain=InProcessToolchain, LaunchCount=1, TargetCount=15, WarmupCount=10)
Runtime = ; GC =
Mean = 5.6903 ns, StdErr = 0.0193 ns (0.34%); N = 15, StdDev = 0.0747 ns
Min = 5.5570 ns, Q1 = 5.6381 ns, Median = 5.7103 ns, Q3 = 5.7320 ns, Max = 5.8272 ns
IQR = 0.0939 ns, LowerFence = 5.4973 ns, UpperFence = 5.8728 ns
ConfidenceInterval = [5.6525 ns; 5.7281 ns] (CI 95%)
Skewness = -0.05, Kurtosis = 2.03


IntroInProcess.'new byte[10kB]': OutOfProc(LaunchCount=1, TargetCount=15, WarmupCount=10)
Runtime = .NET Core 4.6.24628.01, 64bit RyuJIT; GC = Concurrent Workstation
Mean = 438.6340 ns, StdErr = 3.7297 ns (0.85%); N = 15, StdDev = 14.4451 ns
Min = 424.7953 ns, Q1 = 425.9633 ns, Median = 433.1360 ns, Q3 = 449.5959 ns, Max = 469.3894 ns
IQR = 23.6327 ns, LowerFence = 390.5143 ns, UpperFence = 485.0449 ns
ConfidenceInterval = [431.3238 ns; 445.9443 ns] (CI 95%)
Skewness = 0.79, Kurtosis = 2.08


IntroInProcess.'stackalloc byte[10kB]': OutOfProc(LaunchCount=1, TargetCount=15, WarmupCount=10)
Runtime = .NET Core 4.6.24628.01, 64bit RyuJIT; GC = Concurrent Workstation
Mean = 5.7449 ns, StdErr = 0.0895 ns (1.56%); N = 15, StdDev = 0.3465 ns
Min = 5.4198 ns, Q1 = 5.4567 ns, Median = 5.5326 ns, Q3 = 6.1673 ns, Max = 6.3062 ns
IQR = 0.7107 ns, LowerFence = 4.3907 ns, UpperFence = 7.2333 ns
ConfidenceInterval = [5.5696 ns; 5.9203 ns] (CI 95%)
Skewness = 0.6, Kurtosis = 1.52


Total time: 00:01:17 (77.6 sec)

// * Summary *

BenchmarkDotNet=v0.10.3.20170304-develop, OS=Microsoft Windows 10.0.14393
Processor=Intel(R) Core(TM) i7-6700HQ CPU 2.60GHz, ProcessorCount=8
Frequency=2531250 Hz, Resolution=395.0617 ns, Timer=TSC
dotnet cli version=2.0.0-alpha-004853
  [Host]    : .NET Core 4.6.24628.01, 64bit RyuJIT
  OutOfProc : .NET Core 4.6.24628.01, 64bit RyuJIT

LaunchCount=1  TargetCount=15  WarmupCount=10

                  Method |       Job |          Toolchain |        Mean |    StdErr |     StdDev |  Gen 0 | Allocated |
------------------------ |---------- |------------------- |------------ |---------- |----------- |------- |---------- |
 'stackalloc byte[10kB]' | InProcess | InProcessToolchain |   5.6903 ns | 0.0193 ns |  0.0747 ns |      - |       0 B |
 'stackalloc byte[10kB]' | OutOfProc |            Default |   5.7449 ns | 0.0895 ns |  0.3465 ns |      - |       0 B |
        'new byte[10kB]' | OutOfProc |            Default | 438.6340 ns | 3.7297 ns | 14.4451 ns | 3.1590 |  10.02 kB |
        'new byte[10kB]' | InProcess | InProcessToolchain | 533.4784 ns | 6.2640 ns | 24.2602 ns | 3.1574 |  10.02 kB |
```
Why do we have a huge difference for 'new byte[10kB]' in `InProcess` and `OutOfProc`?

adamsitnik commented:
>  I think that you hurry to do it, I still have some concerns

I am sorry, I did not know that. I can revert the merge commit if you want.

> Why do we have a huge difference for 'new byte[10kB]' in InProcess and OutOfProc?

Do other benchmarks have also such huge difference?

AndreyAkinshin commented:
And some other minor concerns:
1. InProcess benchmark has empty `Runtime` and `GC` values in the "Detailed results" section. They should be filled by the host values.
2. We should print `InProcess` instead of `InProcessToolchain` in the Toolchain columns of the summary table.
3. Eventually, I want to transform `BenchmarkDotNet.Samples` into a gallery of benchmark which teach people how to use our API. `IntroInProcessWrongEnv` shows how to get an error message. I think it should be an integration test instead which verifies that we print this message.
4. It makes sense to creaate a nice attribute for in-process benchmark. For example, we can introduce `InProcessJobAttribute` which is based on `SimpleJobAttribute` and always uses InProcessToolchain. So, we will be able to declare such benchmarks with a single attribute line without of a ManualConfig requirement.
5. It would be also nice to add some info in our documentation which explains the difference between default toolchain and InProcessToolchain and shows how to use it.

AndreyAkinshin commented:
> I am sorry, I did not know that. I can revert the merge commit if you want.

There's nothing so terrible in this merge, we can try to fix it in master. A note for a future: it's ok to quickly merge simple PRs, but huge PRs which contains a lot of tricky logic should be carefully tests for a merge.

> Do other benchmarks have also such huge difference?

I hadn't a lot of time for testing, but it seems that CPU-bound benchmarks work fine. I guess that the problem is about huge memory traffic, maybe it's a GC-related issue. So, we have to look at all the GC settings and check if there is a difference between Default and InProcess toolchains.

ig-sinicyn commented:
@AndreyAkinshin 
> Why do we have a huge difference 

I'm almost sure It's because of GC. More stuff to collect => more time spent in GC. Sadly, there's no diagnoser for % GC time and I cannot proof it easily

AndreyAkinshin commented:
@ig-sinicyn, it seems that we still have the same GC-related problem in `IntroInProcess`.
